### PR TITLE
servicemesh: add make target for local testing

### PR DESCRIPTION
**Pull request by marseel**
Currently, we didn't have an easy way to run ingress/gatewayapi locally to test it. Let's add new `make kind-servicemesh-install-cilium` target that enables both gatewayapi and ingress for easier development and testing locally.

Usage:
```
make kind && make kind-image && make kind-servicemesh-install-cilium
```

### CODE REVIEW

**Review by bimmlerd — APPROVED** [review]
On a meta note, I think Makefile.kind and kind.sh has reached a level of complexity where it should probably be a Go program which allows you to spin up a kind cluster with whatever config you want :thinking:

**Comment by marseel** [comment]
> On a meta note, I think Makefile.kind and kind.sh has reached a level of complexity where it should probably be a Go program which allows you to spin up a kind cluster with whatever config you want 🤔

Or clusters for clustermesh :sweat_smile: 

I agree it would be nice to have something better than Makefile, I'm not sure if the Go program would be easier though.
Maybe, spinning up clusters + loading images as Makefile, and installing Cilium with whatever features as a bash script :thinking:

**Comment by bimmlerd** [comment]
> > On a meta note, I think Makefile.kind and kind.sh has reached a level of complexity where it should probably be a Go program which allows you to spin up a kind cluster with whatever config you want 🤔
> 
> Or clusters for clustermesh 😅
> 
> I agree it would be nice to have something better than Makefile, I'm not sure if the Go program would be easier though. Maybe, spinning up clusters + loading images as Makefile, and installing Cilium with whatever features as a bash script 🤔

I've just spent some time thinking about this with Fischi, I might come up with design doc at some point - for now, this is fine.

File: Makefile.kind
Comment (sayboras) [inline Makefile.kind]: Can you update this command as per below? I recently fix one issue to handle v1.1.0-rc1 pattern.

https://github.com/cilium/cilium/pull/35189

**Review by sayboras — APPROVED** [review]
I have one comment as per below, the rest looks good. 

Thanks a lot for fomalizing this local dev setup.

File: Makefile.kind
Comment (marseel) [reply Makefile.kind]: Thanks Tam, fixed!

**Comment by joestringer** [comment]
> On a meta note, I think Makefile.kind and kind.sh has reached a level of complexity where it should probably be a Go program which allows you to spin up a kind cluster with whatever config you want

Whatever tooling we think is broadly useful to different contributors we can integrate, but personally I find that the more we diverge from `make`, `make build`, `make test` into custom tooling, the harder it is to understand what the different tools do. I like that all I need to know is that you start by running `make help` and then go from there. Even for a lot of the Go tooling, I forget the various extra arguments we need to pass to filter the right directories to test or run a benchmark. We can combat this with good documentation of course. Maybe what I'm saying is "yes, and": As long as we aren't compromising the simple "make" workflows then by all means we can create tailored tooling for this.

File: Makefile.kind
Comment (joestringer) [inline Makefile.kind:315]: nit: IIRC when you define a variable like this in Make, by default it will evaluate the statement every time it is referenced. Given that the GW_VERSION is referenced several times in the make target below, this can cause a bunch of extra shelling out to re-fetch the same information. I believe it's a little more efficient to do this:

```suggestion
DEFAULT_GW_VERSION := $(shell grep "sigs.k8s.io/gateway-api" go.mod | awk '{print $$2}' | awk -F'-' '{print (NF>2)?$$NF:$$0}')
GW_VERSION ?= $(DEFAULT_GW_VERSION)
```

(more important for GW_VERSION than the others given the reuse of the variable, but it's probably not bad style to define them all this way).

File: contrib/testing/kind-servicemesh.yaml
Comment (joestringer) [inline contrib/testing/kind-servicemesh.yaml:1]: nit: We could alternatively put this under `examples/servicemesh/` if we think that this configuration would be commonly used by users as well.

**Review by joestringer — APPROVED** [review]
LGTM for build. Couple of minor optimizations/suggestions.

File: Makefile.kind
Comment (joestringer) [reply Makefile.kind:315]: (I guess the flip side is that this means the DEFAULT_GW_VERSION is *always* evaluated in the Makefile even for developers who never use the target)

Comment (marseel) [reply Makefile.kind:315]: > nit: IIRC when you define a variable like this in Make, by default it will evaluate the statement every time it is referenced. Given that the GW_VERSION is referenced several times in the make target below, this can cause a bunch of extra shelling out to re-fetch the same information. I believe it's a little more efficient to do this:

> (I guess the flip side is that this means the DEFAULT_GW_VERSION is always evaluated in the Makefile even for developers who never use the target)

yes, I thought about it when I wrote it and I did it this way on purpose. In case this command breaks somehow, we don't want everyone to not be able to run other targets / fail building images etc.

File: contrib/testing/kind-servicemesh.yaml
Comment (marseel) [reply contrib/testing/kind-servicemesh.yaml:1]: I'm not sure if that would be a common configuration, essentially it's one of the possibility in https://docs.cilium.io/en/stable/network/lb-ipam/#loadbalancer-ip-address-management-lb-ipam
that we use for testing / local development. 
Usually, I think it would be either a cloud provider handling LB service or BGP AFAIK.

File: Makefile.kind
Comment (joestringer) [reply Makefile.kind:315]: That actually seems ideal :) if something breaks, find out as part of the PR that breaks it rather than subtly breaking something that is more rarely used. But :shrug: I don't feel strongly about this.

**Comment by marseel** [comment]
I marked it as ready-to-merge, no idea why maintainer's little helper didn't kick in. 
I have a review from build, contributing and also servicemesh teams.

**Pull request merged**
